src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs

```diff
@@ -0,0 +1,677 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    ///   Represents a numeric JSON value.
+    /// </summary>
+    public class JsonNumber : JsonNode, IEquatable<JsonNumber>
+    {
+        private string _value;
```

Comment:
This means that constructing a JsonNumber from a value requires converting it to a string?  That seems unnecessarily expensive.  Did we consider alternative representations?

Reply:
We are thinking about changing it into one of the following: 
    -`Span<byte>` or array of `Utf8String`/`Char8` (once they come online in the future) / `byte`  
    - internal types that inherit from `JsonNumber` 
but we left the decision for API review.

Reply:
As @kasiabulat mentioned, we had discussed a few other alternatives and they are generally implementation detail specific. One of the design goals mentioned in the spec (https://github.com/dotnet/corefx/pull/39954) is

> Higher emphasis on usability over allocations/performance.

- Internal types that are specific to each numeric type in .NET with factories to create JsonNumber (@JamesNK suggested - https://github.com/dotnet/corefx/pull/39954#discussion_r309889640)
- Change to a internal struct field which has all the supported numeric types (something @terrajobst suggested that might be worth exploring).

Granted, I agree we should avoid any unnecessary overhead especially if there is no/little regression to usability. **We landed on keeping it as is for now, and revisit this once more of the feature is implemented, to unblock forward progress.**

@kasiabulat - can you please file an issue to explicitly track this? Or update the existing issue to talk about the underlying representation of a JsonNumber rather than just what `SetFormattedValue` does: https://github.com/dotnet/corefx/issues/39957

Another design point was about immutable vs mutable, but we landed on keeping it mutable.

Reply:
> we had discussed a few other alternatives

There are other approaches as well.  For example, rather than having a single string field, it could store:
```C#
private TypeCode _typeOf8bytes;
private ulong _8bytes;
private string _stringValue;
```
This would make the JsonNumber itself a little bigger, but it would allow any of the primitive types that are <= 8 bytes long (which is everything except for `decimal`) to be stored in the `_8bytes` field, with no additional allocation, formatting, or parsing required.  And as long as we expect the majority of `JsonNumber`s created to be created for `byte`s, `sbyte`s, `ushort`s, `short`s, `uint`s, `int`s, `ulong`s, `long`s, `float`s, or `double`s (rather than `decimal`s or `string`s), this would be seem to be a win.

> and they are generally implementation detail specific

Sure... but this PR is providing the implementation :)

> We landed on keeping it as is for now, and revisit this once more of the feature is implemented, to unblock forward progress.

Ok, as long as there's an issue tracking this and we'll revisit it.

Reply:
There was also some discussion on adding BigInteger support, but that would fall into the decimal/string category in the approach you suggested.

> primitive types that are <= 8 bytes long

Hmm, interesting. float/double can be represented as ulong without loss of precision?

Reply:
(I edited the issue: https://github.com/dotnet/corefx/issues/39957)

Reply:
> float/double can be represented as ulong without loss of precision? 

Yes. Don't think of it as a long, think of it as an opaque 8 bytes that can be reinterpreted however you like, e.g. MemoryMarshal.Cast.

Reply:
> Don't think of it as a long, think of it as an opaque 8 bytes that can be reinterpreted however you like

Ah, makes sense (especially now that I look at the implementation of Single/Double more closely). Looks like each possible value in the set of floats/doubles can be uniquely represented by a unique 8-byte sequence.

https://docs.microsoft.com/en-us/dotnet/api/system.double?view=netframework-4.8#floating-point-representation-and-precision
The Double data type stores double-precision floating-point values in a **64-bit binary format**